File: code/56.merge-intervals.go
/*
 * @lc app=leetcode id=56 lang=golang
 *
 * [56] Merge Intervals
 */

// @lc code=start
func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	res, start, end := make([][]int, 0), intervals[0][0], intervals[0][1]

	for i := 0; i < len(intervals); i++ {
		left, right := intervals[i][0], intervals[i][1]
		if left > end {
			res = append(res, []int{start, end})
			start = left
		}

		end = max(end, right)
	}
	res = append(res, []int{start, end})
	return res
}

func max(a, b int) int {
	if a < b {
		a = b
	}
	return a
}

// @lc code=end

File: code/56.merge-intervals.java
/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> list = new ArrayList<>();
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);

        int start = intervals[0][0];
        int end = intervals[0][1];

        for (int i = 1; i < intervals.length; i++) {
            int left = intervals[i][0];
            int right = intervals[i][1];

            if (left > end) {
                list.add(new int[] {start, end});
                start = left;
            }

            end = Math.max(end, right);
        }
        
        list.add(new int[] {start, end});
        return list.toArray(new int[list.size()][]);
    }
}
// @lc code=end

